rfcs/20230704-dynamism-101.md

```diff
@@ -0,0 +1,825 @@
+step. This would allow to roll out the new functionality (e.g.
+`DynamicReduceWindowOp`) sooner, without blocking on refactoring of the
+existing functionality (e.g. updating `ReduceWindowOp` as discussed above).
+
+<!-- markdownlint-disable line-length -->
+## <a id="p6" /> (P6) Represent shape computations as StableHLO operations working on 0-dimensional tensors
+<!-- markdownlint-enable line-length -->
+
+**Baseline:** The StableHLO dialect represents sizes as 1-dimensional tensors,
+with static sizes expressed as `DenseIntElementsAttr` and dynamic sizes
+expressed as `1DTensorOf<[HLO_DimensionValue]>`.
+
+```c++
+class PadOp ... {
+  // This is how static sizes are represented in TableGen:
+  // I64ElementsAttr:$edge_padding_low
+  // I64ElementsAttr:$edge_padding_high
+  // I64ElementsAttr:$interior_padding
+  DenseIntElementsAttr getEdgePaddingLow();
+  DenseIntElementsAttr getEdgePaddingHigh();
+  DenseIntElementsAttr getInteriorPadding();
+}
+
+class DynamicPadOp ... {
+  // This is how dynamic sizes are represented in TableGen:
+  // HLO_DimensionTensor:$edge_padding_low
+  // HLO_DimensionTensor:$edge_padding_high
+  // HLO_DimensionTensor:$interior_padding
+  TypedValue<RankedTensorType> getEdgePaddingLow();
+  TypedValue<RankedTensorType> getEdgePaddingHigh();
+  TypedValue<RankedTensorType> getInteriorPadding();
+}
+```
+
+As mentioned in [F5](#f5), shape computations in StableHLO programs produce
+these 1-dimensional tensors using one of several approaches, including ops from
+Arith, Shape, StableHLO as well as other dialects. E.g. addition of dimension
+sizes can be represented via `stablehlo::AddOp`, `arith::AddIOp`, `shape::AddOp`
+and in a few other ways.
+
+Furthermore, as mentioned in [F4](#f4), using 1-dimensional tensors introduces
+considerable complexity because typical shape computations operate in terms of
+scalars. As a result, in code that constructs dynamic StableHLO ops, it is not
+unusual to encounter computations on 0-dimensional tensors whose results are
+then reshaped to 1-dimensional tensors and then concatenated together
+([example](https://github.com/tensorflow/mlir-hlo/blob/1fd13ef28d3f423363e59b1a80eb2c26e0c0979d/mhlo/transforms/chlo_legalize_to_hlo/chlo_legalize_to_hlo.cc#L1577-L1587)).
+
+**Proposal:** The RFC proposes to standardize on representing shape computations
+as StableHLO operations working on 0-dimensional tensors. More specifically,
+the proposal is to:
+
+1) Affirm that only shape computations that use StableHLO operations are
+   supported in StableHLO portable artifacts (this can be changed in future
+   RFCs, but is out of scope for this one).
+2) Use `Variadic<0DTensorOf<HLO_DimensionValue>>` instead of
+   `HLO_DimensionTensor` in the StableHLO dialect (this is consistent with
+   "variadic number of 0-dimensional tensors of integer type" from [P5](#p5)).
+3) Drop support for unranked dynamism in StableHLO (otherwise the operations
```

Comment:
Is this a minor change in terms of impact? I'm thinking this should maybe be mentioned in the P6 header or elsewhere that is more visible to ensure it doesn't go unnoticed.

Reply:
"This mentions the only use of this as TF, are we confident it isn't used externally?". The RFC mentions that we don't have 100% confidence: "the only user of unranked dynamism in StableHLO/MHLO appears to be TensorFlow", so we definitely need to validate this.

"Or is this a call to ask if any StableHLO/OpenXLA users support unranked dynamism?". Yes, exactly. The hope is that the uses that we haven't accounted for (if any) will come up during the review of the RFC. I've also updated the title of P6 to explicitly call out that the proposal is to drop support for unranked dynamism - that's a good idea.